File: Algo/LeadToOffer/34_FindOnlyOneChar.cpp
// 1. Use map. Store the first position. O(n) time, O(1) space.
class Solution {
public:
    int FirstNotRepeatingChar(string str) {
        if (str.length()==0)
            return -1;
        unordered_map<char, int> dict;
        for (int i=0; i<str.length(); i++) {
            if (dict.find(str[i])==dict.end())
                dict[str[i]]=i;
            else
                dict[str[i]]=-1;
        }
        unordered_map<char,int>::iterator iter;
        int min_pos = INT_MAX;
        for (iter=dict.begin(); iter!=dict.end(); iter++) {
            if (iter->second != -1 && iter->second<min_pos)
                min_pos = iter->second;
        }
        return min_pos<INT_MAX?min_pos:-1;
    }
};

// 2. Use array as map to store at most 256 char.
// The value at pos is the appreance time. O(1) space, O(n) time.
class Solution {
public:
    int FirstNotRepeatingChar(string str) {
        if (str.length()==0)
            return -1;
        unsigned int table[256];
        for (int i=0; i<256; i++) {
            table[i]=0;
        }
        for (int i=0; i<str.length(); i++) {
            table[int(str[i])]++;
        }
        for (int i=0; i<str.length(); i++) {
            if (table[int(str[i])]==1)
                return i;
        }
        return -1;
    }
};

File: Algo/LeadToOffer/34_FindOnlyOneChar.py
## 1. Use map. Store the appearance time. O(n) time, O(1) space.
class Solution:
    def FirstNotRepeatingChar(self, s):
        if len(s)==0:
            return -1
        dict_v = {}
        for c in s:
            if c in dict_v:
                dict_v[c]+=1
            else:
                dict_v[c]=1
        for i, c in enumerate(s):
            if c in dict_v and dict_v[c]==1:
                return i
        return -1

# more clear
class Solution(object):
    def firstUniqChar(self, s):
        """
        :type s: str
        :rtype: str
        """

        d = {}
        for n in s:
            d[n] = d.get(n, 0) + 1

        for n in s:
            if d[n] == 1:
                return n

        return " "